designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/loader/ProcessingTypesConfigBasedProcessingTypeDataLoader.scala

```diff
@@ -5,16 +5,50 @@ import com.typesafe.scalalogging.LazyLogging
 import pl.touk.nussknacker.engine._
 import pl.touk.nussknacker.engine.api.process.ProcessingType
 import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
-import pl.touk.nussknacker.engine.util.loader.ScalaServiceLoader
+import pl.touk.nussknacker.engine.util.loader.{ModelClassLoader, ScalaServiceLoader}
 import pl.touk.nussknacker.ui.configloader.{ProcessingTypeConfigs, ProcessingTypeConfigsLoader}
 import pl.touk.nussknacker.ui.process.processingtype._
 import pl.touk.nussknacker.ui.process.processingtype.loader.ProcessingTypeDataLoader.toValueWithRestriction
 import pl.touk.nussknacker.ui.process.processingtype.provider.ProcessingTypeDataState
 
+import java.nio.file.Path
+import java.util.concurrent.atomic.AtomicReference
+
 class ProcessingTypesConfigBasedProcessingTypeDataLoader(processingTypeConfigsLoader: ProcessingTypeConfigsLoader)
     extends ProcessingTypeDataLoader
     with LazyLogging {
 
+  private object ModelClassloadersHolder {
+
+    private val modelClassloaders: AtomicReference[Option[Map[String, ModelClassLoader]]] =
```

Comment:
Why do you need Option here?

Reply:
To have a more explicit uninitialized state - None instead of empty map. Both approaches are imo ok since in this case I think empty map always means uninitialized state, but I'm subjectively leaning towards Option approach because of its explicitness. Do you think simply Map would be better?

Reply:
Why didn't you create this map once before the rest of the initialization code?  Take a look that we already have some initialization steps and each step depends on other steps which is visible in order of initialization - see `NussknackerAppFactory`

Reply:
In `NussknackerAppFactory` we don't have access to `workingDirectoryOpt` but now I see its for test purposes and during execution its hardcoded to None on https://github.com/TouK/nussknacker/blob/aecd8a0a42aa7fa64ef675697208d0ac3c5c543e/designer/server/src/main/scala/pl/touk/nussknacker/ui/server/AkkaHttpBasedRouteProvider.scala#L774

I'll move it to `NussknackerAppFactory` and hardcode the `workingDirectoryOpt` on that level, thanks :+1: